### ISSUE TITLE
Hashing was accidentally changed.

### ISSUE THREAD

**Issue by cdsmith** [google/codeworld#704]
It looks like on July 20th, hashes changed somehow for compiled programs.  The result is that old deploy links don't work any longer, since they rely on building to write the build artifact with the same hash.  Old source links do work, however, because they load the source, which is indexed by the old hash, and then recompute the hash during the build.

**Comment by cdsmith** [comment]
Culprit is probably https://github.com/google/codeworld/commit/465cf1a1529a78de97d56e63a5416a4a855a5f23

Still figuring out how

**Comment by cdsmith** [comment]
Yes, the timing makes it abundantly clear that this is exactly when it happened.  But I've looked over that commit, and I still can't for the life of me figure out why.  Unless... loading a file from an NFS volume instead of local disk couldn't possibly change its contents somehow... could it?

**Comment by cdsmith** [comment]
Reproducible, for example, by following links from https://www.joachim-breitner.de/blog/737-Reservoir_sampling_with_few_random_bits

At this point, we will need to fix in a way that makes either hashing scheme work for older programs.

**Comment by cdsmith** [comment]
@nomeata FYI, this affects your stuff more than anything else.

**Comment by cdsmith** [comment]
Still no idea why the hashes changed, but I think I have a work-around, by just creating any missing parent directories needed by `codeworld-compiler` for its output.  It turns out `runJS` doesn't need the hash to be consistent for any other reason than because the output files might be in non-existent directories.

**Comment by nomeata** [comment]
Thanks for keeping the old links working, it’s appreciated!

**Comment by cdsmith** [comment]
Following up on this, I actually don't know when hashes changed.  I initially blamed the incorrect has for the problem, but it has nothing to do with it.

The problem was just that I wasn't creating the output directories in the handler for /runJS.  Note that in `codeworld-compiler:Main`, `compileHandler` performs a `ensureBuildDir`, but `runHandler` doesn't.  That means that when you build code from the CodeWorld environment, it works, and the link to share the built code initially works; but if the build directory is deleted later, the share link breaks, because it doesn't have the expected directory in which to write the build result.

This started on July 20th because I moved build output, which previously lived alongside the source, into its own separate tree. When build outputs were stored next to the source, they always existed for previously built code.  But when they moved into a separate directory tree, the source directory may already exist, but the build output dir may not.  Hence, the bug.

This was fixed correctly by having `codeworld-compiler` create the necessary output locations.  In fact, I should remove `ensureBuildDir` from the code; it should not be necessary.

The changed hashes probably happened at a whole different time.  The change is interesting, but it turns out we don't rehash the code stored at the location of the old hash; so it still works fine, even if the hash is technically wrong now.  I could write whatever code I wanted to whatever program id I wanted, behind the scenes, and it would be correctly servable and runnable.  I like that.

I'll declare this fixed, and not worry about the hash change.  I did open a couple bugs about monitoring improvements, as a follow-up (#705 and #706).
